src/shared/GitHub/GitHubAuthentication.cs

```diff
@@ -27,50 +59,107 @@ public class GitHubAuthentication : AuthenticationBase, IGitHubAuthentication
+                var promptArgs = new StringBuilder("prompt");
+                if ((modes & AuthenticationModes.Basic) != 0) promptArgs.Append(" --basic");
+                if ((modes & AuthenticationModes.OAuth) != 0) promptArgs.Append(" --oauth");
 
-                if (!resultDict.TryGetValue("username", out userName))
+                IDictionary<string, string> resultDict = await InvokeHelperAsync(helperPath, promptArgs.ToString(), null);
+
+                if (!resultDict.TryGetValue("mode", out string responseMode))
                 {
-                    throw new Exception("Missing username in response");
+                    throw new Exception("Missing 'mode' in response");
                 }
 
-                if (!resultDict.TryGetValue("password", out password))
+                switch (responseMode.ToLowerInvariant())
                 {
-                    throw new Exception("Missing password in response");
+                    case "oauth":
+                        return new AuthenticationPromptResult(AuthenticationModes.OAuth);
+
+                    case "basic":
+                        if (!resultDict.TryGetValue("username", out string userName))
+                        {
+                            throw new Exception("Missing 'username' in response");
+                        }
+
+                        if (!resultDict.TryGetValue("password", out string password))
+                        {
+                            throw new Exception("Missing 'password' in response");
+                        }
+
+                        return new AuthenticationPromptResult(new GitCredential(userName, password));
+
+                    default:
+                        throw new Exception($"Unknown mode value in response '{responseMode}'");
                 }
             }
             else
             {
                 ThrowIfTerminalPromptsDisabled();
 
-                Context.Terminal.WriteLine("Enter GitHub credentials for '{0}'...", targetUri);
+                switch (modes)
+                {
+                    case AuthenticationModes.Basic | AuthenticationModes.OAuth:
+                        var menuTitle = $"Select an authentication method for '{targetUri}'";
+                        var menu = new TerminalMenu(Context.Terminal, menuTitle)
+                        {
+                            new TerminalMenuItem(1, "Web browser", true),
+                            new TerminalMenuItem(2, "Username/password")
+                        };
 
-                userName = Context.Terminal.Prompt("Username");
-                password = Context.Terminal.PromptSecret("Password");
-            }
+                        int option = menu.Show();
+
+                        if (option == 1) goto case AuthenticationModes.OAuth;
+                        if (option == 2) goto case AuthenticationModes.Basic;
```

Comment:
This deviation from normal style is probably worth it here.

Reply:
Yeah. I'm usually quite strict about 'one liner' control flow blocks, but this form seemed to improve clarity IMO.